rfc/canonical-endpoint-dns.md

```diff
@@ -0,0 +1,76 @@
+This hostname can be used as a default endpoint for Constellation for interactions with the bootstrapper, cluster joining and the kubernetes API.
+
+Additionally, customers may register custom domains and point them to the loadbalancer hostname.
+
+### Azure
+
+On Azure, each public ip [can have a public dns name](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/public-ip-addresses#dns-name-label) in the format `name-1234567890.location.cloudapp.azure.com`.
+This hostname can be used as a default endpoint for Constellation for interactions with the bootstrapper, cluster joining and the kubernetes API.
+
+Additionally, customers may register custom domains and point them to either the public ip hostname or address.
+
+### GCP
+
+On Google Cloud, neither loadbalancers nor public ips have a default, public dns name.
+
+We have to use the (single) public ipv4 address as the default endpoint for Constellation for interactions with the bootstrapper, cluster joining and the kubernetes API.
+
+Additionally, customers may register custom domains and point them to the public ip address.
+
+### Conclusion
+
+**Fallback endpoint**:
+
+- loadbalancer dns name on AWS (`name-1234567890.region.elb.amazonaws.com`)
+- public ip dns name on Azure (`name-1234567890.location.cloudapp.azure.com`)
+- public ip on GCP (`203.0.113.1`)
+
+**(Optional) customer provided dns**:
+
+- any custom domain that points to the fallback endpoint
+
+## Steps for enabling DNS (aka: "canonical endpoint")
+
+1. start adding the fallback endpoint (and the custom domain, if set) to the SAN field of the kube-apiserver certificates
+  - SAN should still contain everything it contains before
+    - legacy public ip we use before switching to dns
+    - `kubernetes`, `kubernetes.default`, `kubernetes.default.svc`, `kubernetes.default.svc.cluster.local`
+    - 127.0.0.1, ::1, localhost
+  - fallback endpoint (as specified above)
+  - optional customer provided hostname (if set)
+2. once every apiserver uses the extended SAN field, the cluster-internal advertise-address for the apiserver can be switched to the fallback endpoint
+  - no-op on GCP (this is already the status quo)
+  - ensure every existing kubelet uses the new endpoint
+  - ensure newly joining kubelet use the new endpoint
+  - k8s configmap uses new endpoint (`kube-system/kubeadm-config` -> `.data.ClusterConfiguration`)
+3. switch id file `constellation-id.json` to use either fallback endpoint or customer endpoint (Q: which is better if both are available?)
+4. switch setting in kubernetes config (`constellation-admin.conf`) to use either fallback endpoint or customer endpoint (Q: which is better if both are available?)
+  - in existing clusters: patch config
+  - in new clusters: write correct endpoint on `constellation init`
```

Comment:
```suggestion
1. start adding the fallback endpoint (and the custom domain, if set) to the SAN field of the kube-apiserver certificates
    - SAN should still contain everything it contains before
      - legacy public ip we use before switching to dns
      - `kubernetes`, `kubernetes.default`, `kubernetes.default.svc`, `kubernetes.default.svc.cluster.local`
      - 127.0.0.1, ::1, localhost
    - fallback endpoint (as specified above)
    - optional customer provided hostname (if set)
2. once every apiserver uses the extended SAN field, the cluster-internal advertise-address for the apiserver can be switched to the fallback endpoint
    - no-op on GCP (this is already the status quo)
    - ensure every existing kubelet uses the new endpoint
    - ensure newly joining kubelet use the new endpoint
    - k8s configmap uses new endpoint (`kube-system/kubeadm-config` -> `.data.ClusterConfiguration`)
3. switch id file `constellation-id.json` to use either fallback endpoint or customer endpoint (Q: which is better if both are available?)
4. switch setting in kubernetes config (`constellation-admin.conf`) to use either fallback endpoint or customer endpoint (Q: which is better if both are available?)
    - in existing clusters: patch config
    - in new clusters: write correct endpoint on `constellation init`
```
nit: you need to add extra indentation, otherwise the markdown rendering of the nested lists is broken (see GitHub preview of the file)

Reply:
As for the question about whether to use fallback or custom endpoints for kube-config/id-file, I would suggest using the fallback, as this should be considered the "safer" endpoint.
As this is not user controlled, it should be more resilient to user error as well.
It is unlikely a CSP accidentally removes the default DNS name of a loadbalancer, while I can see a user doing so.